### ISSUE TITLE
Promise.all running promises with async tasks should fire single changedetection

### ISSUE THREAD

**Issue by pamtbaau** [angular/zone.js#1007]
Angular zone fires changedetection when a Promise executes an async task. 

However, when multiple Promises are being run inside Promise.all(), again every single Promise causes changedetection to be fired.

When running the following code:
```
import { Component } from '@angular/core';

@Component({
    selector: 'app-root',
    template: `<p>{{ title }}!</p>
               <button (click)="onClick()">run</button>`,
})
export class AppComponent {
    changeDetectionCount = 0;
    get title(): string {
        this.changeDetectionCount++;
        return 'app';
    }

public onClick() {
    const promises: Array<Promise<void>> = [];
    const loopCount = 10;

    for (let i = 0; i < loopCount; i++) {
        promises.push(new Promise<void>((resolve) => {
            setTimeout(() => {
                resolve();
            }, 0);
        }));
    }
    this.changeDetectionCount = 0;
    Promise.all(promises)
        .then(() => {
            console.log(
                `number of Promises executed: ${loopCount}\n`,
                `changeDetection count: ${this.changeDetectionCount}`
            );
        });
    }
}
```
the following is written to the console: (Number of cycles is 10 in production mode)
```
number of Promises executed: 10
changeDetection count: 20
```
Shouldn't `Promise.all()` silence all inside promises and fire a single changedetection cycle once resolved?

**Comment by JiaLiPassion** [comment]
the `changeDetection` increase because you run `setTimeout`, which will trigger `ngZone.checkStable`.

**Comment by pamtbaau** [comment]
It will also be triggered by any other async task inside the Promises. For example Node's fs.exists() (in Electron app), or insert/update/deletes in SQLite in Electron.

I found out that 7500 inserts in SQLite resulted in in 7500+ changedetection cycles.

I know that `setTimeout` will trigger ngZone, but within a `Promise.all()` it doesn't make sense to fire it for each and every promise containing an async task. `Promise.all()` wraps all contained promises into a single statement with a single `.then()`.  Shouldn't `Promise.all()`, fire `ngZone.checkStable` only once for all contained promises combined?

**Comment by JiaLiPassion** [comment]
the reason is every `promise` it self will also run in `ng_zone`, I understand your point. 
current behavior is:

```javascript
const promise1 = Promise.resolve(1).then(() => {
  this.text1 = 'promise1 resolved`; // trigger CD
});

const promise2 = Promise.resolve(2).then(() => {
  this.text2 = 'promise2 resolved'; // trigger CD again
});

Promise.all([promise1, promise2]).then(() => {
  this.text3 = 'all promises resolved'; // finally trigger CD here.
});
```

I am not sure what is the correct behavior, @mhevery , could you check this one? thank you.

**Comment by mhevery** [comment]
The current mental model is that invocation of `.then` is when we capture the zone. Hence `then`'s function will run in the captured zone. `Promise.all` calls `.then` on `promise1` and `promise2` internally. So if you want to do what you are asking you should do it this way.

```
// assuming in AngularZone here
const promise1 = new Promise((res, res) => {
   setTimeout(res, 300);
}) ;

const promise2 = new Promise((res, res) => {
   setTimeout(res, 500);
}) ;

outsideAngularZone.run(() => Promise.all([promise1, promise2])).then(() => {
  // promise1.then and promise2.then will also trigger change detection twice.
  this.text3 = 'all promises resolved'; // finally trigger change detection here.
});
```

**Comment by pamtbaau** [comment]
@mhevery , Thanks for the explaination. Will get back on this after doing some tests tomorrow...

**Comment by pamtbaau** [comment]
As promised, some tests.

TL;DR:
1. No ngZone. runOutsideAngular() : 
**Each promise triggers change detection**
2. Create promises outside runOutsideAngular, run  Promise.all() inside:  
**Each promise triggers change detection**
3. Create promises and run promises inside runOutsideAngular: 
**No more change detection triggered by promises**

None of the above scenario's are desirable IMHO. When using Promise.all(), I tell the system not to get back before each and every promise has been resolved. Only then I would expect change detection

Test results:

1. No ngZone. runOutsideAngular() :
```
import { Component, NgZone } from '@angular/core';

@Component({
    selector: 'app-root',
    template: `
        {{title}}
        <button (click)="onClick()">Run</button>
        `,
})
export class AppComponent {
    public get title(): string {
        console.log('> detection');
        return 'app';
    }

    constructor(private zone: NgZone) { }

    public onClick() {
        console.log('Enter onClick()');
        console.log('Create Promise 1');
        const promise1 = new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve();
            }, 300);
        })
        .then(() => {
            console.log('Promise 1 then');
        });

        console.log('Create Promise 2');
        const promise2 = new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve();
            }, 500);
        })
        .then(() => {
            console.log('Promise 2 then');
        });

        Promise.all([promise1, promise2])
        .then(() => {
            console.log('Promise.all() then');
        });
        console.log('Exit onClick()\n');
    }
}
```
Results:
```
Enter onClick()
Create Promise 1
Create Promise 2
Exit onClick()

> detection
> detection
Promise 1 then
> detection
> detection
Promise 2 then
Promise.all() then
> detection
> detection
```
Conslusion: Detection after onClick + detection for each Promise.

2. Create promises outside runOutsideAngular, run  Promise.all() inside: 

```
    public onClick() {
        console.log('Enter onClick()');
        console.log('Create Promise 1');
        const promise1 = new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve();
            }, 300);
        })
        .then(() => {
            console.log('Promise 1 then');
        });

        console.log('Create Promise 2');
        const promise2 = new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve();
            }, 500);
        })
        .then(() => {
            console.log('Promise 2 then');
        });

        console.log('Before runOutsideAngular()');
        this.zone.runOutsideAngular(() => {
            console.log('Inside runOutsideAngular');
            Promise.all([promise1, promise2])
            .then(() => {
                console.log('Promise.all() then');
            });
        });
        console.log('After runOutsideAngular()');
        console.log('Exit onClick()\n');
    }
}
```
Results:
```
Enter onClick()
Create Promise 1
Create Promise 2
Before runOutsideAngular()
Inside runOutsideAngular
After runOutsideAngular()
Exit onClick()

> detection
> detection
Promise 1 then
> detection
> detection
Promise 2 then
> detection
> detection
Promise.all() then
```
Conclusion: Detection after onClick + detection for each Promise. Same number of detection cycles.

3. Create promises and run promises inside runOutsideAngular: 
```
    public onClick() {
        console.log('Enter onClick()');

        console.log('Before runOutsideAngular()');
        this.zone.runOutsideAngular(() => {
            console.log('Inside runOutsideAngular');

            console.log('Create Promise 1');
            const promise1 = new Promise((resolve, reject) => {
                setTimeout(() => {
                    resolve();
                }, 300);
            })
            .then(() => {
                console.log('Promise 1 then');
            });

            console.log('Create Promise 2');
            const promise2 = new Promise((resolve, reject) => {
                setTimeout(() => {
                    resolve();
                }, 500);
            })
            .then(() => {
                console.log('Promise 2 then');
            });

            Promise.all([promise1, promise2])
            .then(() => {
                console.log('Promise.all() then');
            });
        });
        console.log('After runOutsideAngular()');
        console.log('Exit onClick()\n');
    }
```
Results:
```
Enter onClick()
Before runOutsideAngular()
Inside runOutsideAngular
Create Promise 1
Create Promise 2
After runOutsideAngular()
Exit onClick()

> detection
> detection
Promise 1 then
Promise 2 then
Promise.all() then
```
Conclusion: No more detection caused by Promises.

**Comment by JiaLiPassion** [comment]
@pamtbaau , thank you for the test. you can try this.

```javascript
this.zone.runOutsideOfAngular(() => {
  const promise1 = new Promise((resolve, reject) => {
                setTimeout(() => {
                    resolve();
                }, 300);
            })
            .then(() => {
                console.log('Promise 1 then');
            });
});

this.zone.runOutsideOfAngular(() => {
  const promise2 = new Promise((resolve, reject) => {
                setTimeout(() => {
                    resolve();
                }, 300);
            })
            .then(() => {
                console.log('Promise 2 then');
            });
});

Promise.all([promise1, promise2]).then(() => {
                  console.log('Promise.all() then');
});

```

**Comment by pamtbaau** [comment]
Per your request...

Code:
```
    public onClick() {
        let promise1: Promise<void>;
        let promise2: Promise<void>;

        console.log('Enter onClick()');

        console.log('Before runOutsideAngular()');
        this.zone.runOutsideAngular(() => {
            console.log('Inside runOutsideAngular');

            console.log('Create Promise 1');
            promise1 = new Promise((resolve, reject) => {
                setTimeout(() => {
                    resolve();
                }, 300);
            })
            .then(() => {
                console.log('Promise 1 then');
            });

            console.log('Create Promise 2');
            promise2 = new Promise((resolve, reject) => {
                setTimeout(() => {
                    resolve();
                }, 500);
            })
            .then(() => {
                console.log('Promise 2 then');
            });

        });
        console.log('After runOutsideAngular()');

        console.log('Run Promise.all()');
        Promise.all([promise1, promise2])
            .then(() => {
                console.log('Promise.all() then');
            });

        console.log('Exit onClick()\n');
    }
```
Results:
```
Enter onClick()
Before runOutsideAngular()
Inside runOutsideAngular
Create Promise 1
Create Promise 2
After runOutsideAngular()
Run Promise.all()
Exit onClick()

> detection
> detection
Promise 1 then
> detection
> detection
Promise 2 then
Promise.all() then
> detection
> detection
```

Still not to my liking... :-)

**Comment by JiaLiPassion** [comment]
@pamtbaau , I see, thank you for the test.
@mhevery , should we run the `internal then` which automatically created by `Promise.all` in `root zone`?

**Comment by mhevery** [comment]
- The issue is the `setTimeout` which is why you get CD every time. You need to run the `setTimeout` outside the angular zone. Then it should skip CD. Perhaps create a plunker so we can go back and forth. 

- I don't think it is a good idea to run `Promise.all` in root zone. Because if the zone would be watching outstanding microtasks, it would see it go to 0 and than cause CD. Once it would be resolved it would go to > 0 and when it goes to zero it would cause another CD.

I think the correct solution to the problem is to run `setTimeout`  outside the angular zone and than run `Promise.all` outside angular zone as well.

**Comment by JiaLiPassion** [comment]
@mhevery 
- at first, I also think this the `setTimeout` reason, but in @pamtbaau 's sample, he has 10 `promise` and cause 20 `CD`.
- and I mean run `Promise.all` in the zone which the promise.then is called (ngZone), but run `internal` promise.then of the promise array element in root zone.
below is the `Promise.all` code in `zone.js`.

```javascript
 for (let value of values) {
        if (!isThenable(value)) {
          value = this.resolve(value);
        }
        // ********* I mean run this `value.then` in root zone ********
        value.then(
            ((index) => (value: any) => {
              resolvedValues[index] = value;
              count--;
              if (!count) {
                // ***** and run this final resolve in the zone which Promise.all is called (in this example, ngZone) ********
                resolve(resolvedValues);
              }
            })(count),
            reject);
        count++;
      }
```

I have created `stackblitz`, https://stackblitz.com/edit/angular-aweps9?file=app%2Fapp.component.ts , please review , thank you!

**Comment by pamtbaau** [comment]
@JiaLiPassion Tried to review your stackblitz, but it is way over my head... :-)

@mhevery 
It is a good thing that CD happens when a Promise resolves after the onClick() method has finished. And if separate Promises resolve at different moments in time, multiple CD's would be appropriate. However Promise.all() will bundle all promises and will not resolve before all promises have resolved. A single CD would be a logical behaviour when using Promise.all(). IMHO that is... 

Can we agree that this should be the right behaviour?

Please note that I have taken setTimeout just as and example. It happens with (all?) other async tasks like Node's fs.exists('filename', cb), or sqlite.run('select...', cb) that I have wrapped in Promises.

**Comment by JiaLiPassion** [comment]
@pamtbaau 

in fact there are two things in your example.

- promise in promise array should trigger `CD` separately if it is bundled in to a `Promise.all`. Or just one `CD` triggered by `Promise.all.then`.
- the `async task` such `setTimeout`, `fs.exists` you run in promise, you are using `electron`? 
  - if you use such as `setTimeout`, because `setTimeout` is triggered `CD` which has nothing to do with `Promise.all` issue.
  - if you are using `electron` and import `dist/zone.js/zone-mix`, such as `fs.exists` will also trigger `CD`.

**Comment by pamtbaau** [comment]
Yes, I do use Electron but also cross tested my findings from  Electron in a plain browser project. I do not `import dist/zone.js/zone-mix` however.
Btw. I only use setTimeout in my tests as a replacement for` sqlite.run('select...', cb)`. 

I'm sorry, I'm not sure if I understand the two points you make.... 

Can we agree that it makes sense that multiple promises run inside `Promise.all(promises)` should trigger a single CD?

**Comment by JiaLiPassion** [comment]
@pamtbaau 

The two points I mean is 

```javascript
const promise1 = new Promise((res, rej) => {
  setTimeout(res, 100); // setTimeout will trigger CD
});
```

-Point1:  in the code above, setTimeout will trigger CD , and Promise.then will trigger CD too. 
- Point2: it is your question that should `Promise.all(promises)` only trigger a single CD. I am not sure about this one, I agree trigger 1 CD make more sense, please @mhevery  confirm this one.

**Comment by mhevery** [comment]
Can someone create this in stackblitz to verify the behavior.

# Variant A
```
class MyComponent {
  onClick() {
    // Assume each change detection prints `change-detection`
    // Assume we are in angular zone such as a click handler.
    const joinPromise = ngZone.runOutside(() => {
      const promise1 = new Promise((res, res) => setTimeout(res, 300));
      const promise2 = new Promise((res, res) => setTimeout(res, 500));
      return Promise.all(promise1, promise2);
    });

    joinPromise.then(() => {
      console.log('joinPromiseResolution');
    });
    console.log('click handler');
  }
```

should print:

```
click handler
change-detection
-- wait 500 ms --
joinPromiseResolution
change-detection
```

This I believe is what @pamtbaau is looking for

# Variant B
```
class MyComponent {
  onClick() {
    // Assume each change detection prints `change-detection`
    // Assume we are in angular zone such as a click handler.
    const [promise1, promise2] = ngZone.runOutside(() => {
      const promise1 = new Promise((res, res) => setTimeout(res, 300));
      const promise2 = new Promise((res, res) => setTimeout(res, 500));
      return [promise1, promise2];
    });

    // HERE: `Promise.all` is inside `NgZone`
    const joinPromise = Promise.all(promise1, promise2)

    joinPromise.then(() => {
      console.log('joinPromiseResolution');
    });
    console.log('click handler');
  }
```

should print:

```
click handler
change-detection
-- wait 300 ms --
change-detection    <=== Extra Change detection here
-- wait 200 ms --
joinPromiseResolution
change-detection
```

The extra change detection is because the `Promise.all` invoked `promise1.then` inside `NgZone` so when `promise1` resolved its `then` handler ran inside `NgZone` which caused CD.

**Comment by JiaLiPassion** [comment]
@mhevery , Thank you ! yes, you are right, I have created a stackblitz here, https://stackblitz.com/edit/angular-k1hpi2?file=app%2Fapp.component.ts

**Comment by pamtbaau** [comment]
These are certainly good workarounds. Thank you.

But... (it somehow always gets smelly when there is a but involved) I didn't know sqlite3 module was triggering CD on each insert until I needed to insert 5000+ rows which frooze my app for a few minutes.

I wonder, how do I know upfront I need the workaround when using a library? When Promise.all() would silence all triggers and issue a single trigger when resolved, it would be transparent to me.

**Comment by JiaLiPassion** [comment]
@pamtbaau , I believe `sqlite3` also use some common `async operations` such as `setTimeout` or `Promise`, so it will also in `ngZone` and trigger CD, so you can use the method above to run all `sqlite3` operations `outsideAngular` to avoid CD.

**Comment by mhevery** [comment]
@pamtbaau That is an excellent question, to which there is no simple answer. :-( What you need to do is to make sure that your library runs outside of the zone using hacks as described above. 

A better answer and something we are planning on doing is enable way of writing applications without zone by explicitly scheduling CD in your code.

**Comment by mhevery** [comment]
Also I think we should close this issue as working as intended. Do let me know if that is OK with you. Happy to continue to have discussion.

**Comment by pamtbaau** [comment]
It may not be satisfactory, but it is the way it is... The issue may be closed.
